diff --git a/schema.module b/schema.module
index ecc3788..a54a5c5 100644
--- a/schema.module
+++ b/schema.module
@@ -548,13 +548,31 @@ function schema_compare_table($ref, $inspect = NULL) {
     foreach ($col_keys as $key => $val) {
 
       if (!(
-          // First part tests that item exists for both and has same value in both places
-           (isset($col[$key]) && !is_null($col[$key]) && $col[$key] !== FALSE
-          && isset($inspect['fields'][$colname][$key]) && $inspect['fields'][$colname][$key] !== FALSE
-          && $col[$key] == $inspect['fields'][$colname][$key] )
-          // Second test is that it does not exist or exists but is null in both places
-       || ((!isset($col[$key]) || is_null($col[$key]) || $col[$key] === FALSE)
-             && (!isset($inspect['fields'][$colname][$key]) || $inspect['fields'][$colname][$key] === FALSE) ) ) ) {
+        // First part tests that item exists for both and has same value in both places
+        (
+          isset($col[$key]) &&
+          !is_null($col[$key]) &&
+          $col[$key] !== FALSE &&
+          isset($inspect['fields'][$colname][$key]) &&
+          !empty($inspect['fields'][$colname][$key]) &&
+          $inspect['fields'][$colname][$key] !== FALSE &&
+          $col[$key] == $inspect['fields'][$colname][$key]
+        )
+        ||
+        // Second test is that it does not exist or exists but is null in both places
+        (
+          (
+            !isset($col[$key]) ||
+            is_null($col[$key]) ||
+            $col[$key] === FALSE
+          ) &&
+          (
+            !isset($inspect['fields'][$colname][$key]) ||
+            empty($inspect['fields'][$colname][$key]) ||
+            $inspect['fields'][$colname][$key] === FALSE
+          )
+        )
+      )) {
         // One way or another, difference between the two so note it to explicitly identify it later.
         $kdiffs[] = $key;
       }
