diff --git a/date.module b/date.module
index e667b5d..bb07cdd 100644
--- a/date.module
+++ b/date.module
@@ -651,7 +651,9 @@ function date_entity_metadata_field_getter($entity, array $options, $name, $enti
  * Getter callback to return date values as datestamp in UTC.
  */
 function date_entity_metadata_struct_getter($item, array $options, $name, $type, $info) {
-  $value = trim($item[$name]);
+  if (isset($item[$name])) {
+    $value = trim($item[$name]);
+  }
   if (empty($value)) {
     return NULL;
   }
