diff --git a/profiles/commons/modules/commons/commons_wikis/commons_wikis.features.field_instance.inc b/profiles/commons/modules/commons/commons_wikis/commons_wikis.features.field_instance.inc
index 9923eed..df08d4a 100644
--- a/profiles/commons/modules/commons/commons_wikis/commons_wikis.features.field_instance.inc
+++ b/profiles/commons/modules/commons/commons_wikis/commons_wikis.features.field_instance.inc
@@ -108,50 +108,6 @@ function commons_wikis_field_default_field_instances() {
     ),
   );
 
-  // Exported field_instance: 'node-wiki-title_field'
-  $field_instances['node-wiki-title_field'] = array(
-    'bundle' => 'wiki',
-    'default_value' => NULL,
-    'deleted' => 0,
-    'description' => '',
-    'display' => array(
-      'default' => array(
-        'label' => 'above',
-        'settings' => array(),
-        'type' => 'hidden',
-        'weight' => -50,
-      ),
-      'teaser' => array(
-        'label' => 'above',
-        'settings' => array(),
-        'type' => 'hidden',
-        'weight' => -50,
-      ),
-    ),
-    'display_in_partial_form' => 1,
-    'entity_type' => 'node',
-    'field_name' => 'title_field',
-    'label' => 'Title',
-    'required' => 1,
-    'settings' => array(
-      'hide_label' => array(
-        'entity' => 0,
-        'page' => 0,
-      ),
-      'text_processing' => 0,
-      'user_register_form' => FALSE,
-    ),
-    'widget' => array(
-      'active' => 1,
-      'module' => 'text',
-      'settings' => array(
-        'size' => 60,
-      ),
-      'type' => 'text_textfield',
-      'weight' => -50,
-    ),
-  );
-
   // Translatables
   // Included for use with string extractors like potx.
   t('Body');
diff --git a/profiles/commons/modules/commons/commons_wikis/commons_wikis.info b/profiles/commons/modules/commons/commons_wikis/commons_wikis.info
index 9ed837d..073c142 100644
--- a/profiles/commons/modules/commons/commons_wikis/commons_wikis.info
+++ b/profiles/commons/modules/commons/commons_wikis/commons_wikis.info
@@ -28,7 +28,6 @@ features[ctools][] = views:views_default:3.0
 features[features_api][] = api:2
 features[field_instance][] = message-commons_wikis_wiki_updated-field_target_nodes
 features[field_instance][] = node-wiki-body
-features[field_instance][] = node-wiki-title_field
 features[message_type][] = commons_wikis_wiki_updated
 features[node][] = wiki
 features[og_features_permission][] = node:group:create wiki content
