diff --git a/profiles/commons/modules/commons/commons_groups/commons_groups.info b/profiles/commons/modules/commons/commons_groups/commons_groups.info
index f717801..3e128f0 100644
--- a/profiles/commons/modules/commons/commons_groups/commons_groups.info
+++ b/profiles/commons/modules/commons/commons_groups/commons_groups.info
@@ -46,8 +46,6 @@ features[image][] = 50x50
 features[node][] = group
 features[og_features_permission][] = user:user:subscribe
 features[og_features_permission][] = user:user:subscribe without approval
-features[user_permission][] = create group content
-features[user_permission][] = edit own group content
 features[variable][] = menu_options_group
 features[variable][] = menu_parent_group
 features[variable][] = pathauto_node_group_pattern
