diff --git a/libraries.module b/libraries.module
index 62d3d79..63b5d1a 100644
--- a/libraries.module
+++ b/libraries.module
@@ -576,7 +576,7 @@ function libraries_detect($name = NULL) {
       $library['version'] = call_user_func_array($library['version callback'], array_merge(array($library), $library['version arguments']));
     }
     else {
-      $library['version'] = call_user_func($library['version callback'], $library, $library['version arguments']);
+      $library['version'] = call_user_func_array($library['version callback'], array(&$library, $library['version arguments']));
     }
     if (empty($library['version'])) {
       $library['error'] = 'not detected';
